### PR TITLE
Modernize with clang-tidy from base subprojects [PCL-2731]

### PR DESCRIPTION
**Pull request by gnawme**
Initial modernization of code base using `clang-tidy` and an innocuous subset of its `modernize` checks:

- [ ] modernize-avoid-bind
- [ ] modernize-avoid-c-arrays
- [ ] modernize-concat-nested-namespaces
- [x] modernize-deprecated-headers
- [ ] modernize-deprecated-ios-base-aliases
- [ ] modernize-loop-convert
- [ ] modernize-make-shared
- [ ] modernize-make-unique
- [ ] modernize-pass-by-value
- [ ] modernize-raw-string-literal
- [x] modernize-redundant-void-arg
- [ ] modernize-replace-auto-ptr
- [ ] modernize-replace-random-shuffle
- [ ] modernize-return-braced-init-list
- [ ] modernize-shrink-to-fit
- [ ] modernize-unary-static-assert
- [ ] modernize-use-auto
- [ ] modernize-use-bool-literals
- [x] modernize-use-default-member-init
- [ ] modernize-use-emplace
- [x] modernize-use-equals-default
- [x] modernize-use-equals-delete
- [ ] modernize-use-nodiscard
- [ ] modernize-use-noexcept
- [x] modernize-use-nullptr
- [x] modernize-use-override
- [ ] modernize-use-trailing-return-type
- [ ] modernize-use-transparent-functors
- [ ] modernize-use-uncaught-exceptions
- [x] modernize-use-using

This PR supersedes #4249, which was much too large for sensible review.

It takes the approach of starting at the base of the PCL dependency graph, and only applying `clang-tidy` in those subprojects.

https://user-images.githubusercontent.com/1709142/87282880-fd5cc080-c4f4-11ea-878e-9c7de9c71eb8.png

### CODE REVIEW

**Comment by gnawme** [comment]
@kunaltyagi Per @SergioRAgostinho I replaced #4249 with the first of what will be a series of much smaller PRs to introduce `clang-tidy` improvements into the code base.

**Comment by gnawme** [comment]
@kunaltyagi CI is green for this PR as well

**Comment by Morwenn** [comment]
Shoudn't `CMAKE_EXPORT_COMPILE_COMMANDS` rather be passed as a `-D`command-line option to CMake in `tidy-modern.yml`? Adding them to a bunch of `CMakeLists.txt` seems both redundant and not something you want to do most of the time.

**Comment by gnawme** [comment]
> Shoudn't `CMAKE_EXPORT_COMPILE_COMMANDS` rather be passed as a `-D`command-line option to CMake in `tidy-modern.yml`? Adding them to a bunch of `CMakeLists.txt` seems both redundant and not something you want to do most of the time.

That was done intentionally to limit the scope of `clang-tidy` changes to those projects where the `CMakeLists.txt` was modified. 

For contrast, see #4249 where `CMAKE_EXPORT_COMPILE_COMMANDS` was put into the top-level `CMakeLists.txt`.  Nobody wanted to review that PR.

**Comment by Morwenn** [comment]
Oh, my bad then.

File: io/src/openni_camera/openni_driver.cpp [hunk 1]
```diff
@@ -557,12 +557,7 @@ openni_wrapper::OpenNIDriver::DeviceContext::DeviceContext (const xn::NodeInfo&
...
 openni_wrapper::OpenNIDriver::DeviceContext::DeviceContext (const DeviceContext& other)
-: device_node (other.device_node)
-, image_node (other.image_node)
-, depth_node (other.depth_node)
-, ir_node (other.ir_node)
-, device (other.device)
-{
-}
+ 
+= default;
```
Comment (larshg) [inline io/src/openni_camera/openni_driver.cpp:561]: Will the default behavior set all these fields?

**Comment by larshg** [comment]
The addition of the `tidy-modern.yml` will it make a check like the format checker on the CI's? and stop the rest of the builds to continue or where will it be run/added/shown?

File: io/src/openni_camera/openni_driver.cpp (hunk 1 above)
Comment (gnawme) [reply io/src/openni_camera/openni_driver.cpp:561]: Yes; unlike other linters, `clang-tidy` leverages the clang compiler to perform lexical analysis, so I tend to trust its recommendations. (@kunaltyagi can probably comment about the `pcl::weak_ptr` copy construction behavior...)

**Comment by gnawme** [comment]
> The addition of the `tidy-modern.yml` will it make a check like the format checker on the CI's? and stop the rest of the builds to continue or where will it be run/added/shown?

That's the intent; @kunaltyagi wanted to prevent back-sliding in code that's already been tidied.

**Review by larshg — DISMISSED** [review]

**Comment by larshg** [comment]
It looks fine for me. I'm just wondering, if this action doesn't run succesfully for a PR, would it then "block" the PR? 
Wonder if its ends up as a check?

There is however some difference for the I/O module on latest run on your branch @gnawme. In https://github.com/gnawme/pcl/actions/runs/451637712

```
--- a/io/src/image_depth.cpp
+++ b/io/src/image_depth.cpp
@@ -66,7 +66,7 @@ pcl::io::DepthImage::DepthImage (FrameWrapper::Ptr depth_metadata, float baselin
 
 
 pcl::io::DepthImage::~DepthImage ()
-{}
+= default;
 
 
```

Missing those?

**Comment by gnawme** [comment]
> It looks fine for me. I'm just wondering, if this action doesn't run succesfully for a PR, would it then "block" the PR?
> Wonder if its ends up as a check?
> 
> There is however some difference for the I/O module on latest run on your branch @gnawme. In https://github.com/gnawme/pcl/actions/runs/451637712
> 
> ```
> --- a/io/src/image_depth.cpp
> +++ b/io/src/image_depth.cpp
> @@ -66,7 +66,7 @@ pcl::io::DepthImage::DepthImage (FrameWrapper::Ptr depth_metadata, float baselin
>  
>  
>  pcl::io::DepthImage::~DepthImage ()
> -{}
> += default;
>  
>  
> ```
> 
> Missing those?

@larshg Yes, @SergioRAgostinho asked me to remove unnecessary default constructs/destructors in the previous PR, perhaps I can do that when I add `clang-tidy` to the next subproject(s)?

File: .github/workflows/clang-tidy-modern.sh [hunk 2]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+run-clang-tidy-10.py -config='' -header-filter='^((?!.*kiss_fft\.h.*).)*$,.*' -fix
```
Comment (kunaltyagi) [inline .github/workflows/clang-tidy-modern.sh:2]: Please add a reason for excluding kiss_fft files

File: .clang-tidy [hunk 3]
```diff
@@ -0,0 +1,2 @@
+---
+Checks: '-*,modernize-deprecated-headers,modernize-redundant-void-arg,modernize-replace-random-shuffle,modernize-use-equals-default,modernize-use-equals-delete,modernize-use-nullptr,modernize-use-override,modernize-use-using'
```
Comment (kunaltyagi) [inline .clang-tidy:2]: Would it be possible to have them one per line? If not, then please ignore the comment

File: .github/workflows/tidy-modern.yml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+jobs:
+  tidy:
+    runs-on: ubuntu-latest
+    container:
+      image: pointcloudlibrary/env:18.04
+    
+    steps:
+    - name: Install git
+      run: |
+          add-apt-repository ppa:git-core/ppa
```
Comment (kunaltyagi) [inline .github/workflows/tidy-modern.yml:16]: Any reason for installing git?

File: .github/workflows/tidy-modern.yml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+    - uses: actions/checkout@v2
+
+    - name: Run clang-tidy
+      run: |
+          wget https://apt.llvm.org/llvm.sh
+          chmod +x llvm.sh
+          ./llvm.sh 10
+          apt-get update
+          apt-get install -y clang-tidy-10 clang-tools-10 python-pip
```
Comment (kunaltyagi) [inline .github/workflows/tidy-modern.yml:29]: Might make sense to add step from [this](https://apt.llvm.org/) in the env Dockerfile, and then use env:20.04 which has llvm-10

**Review by kunaltyagi — COMMENTED** [review]
Code changes LGTM. Just a few questions on the surrounding infra.

Apologies, but I had to be away for a while, and have lost the original context of the PR

File: .github/workflows/clang-tidy-modern.sh (hunk 2 above)
Comment (gnawme) [reply .github/workflows/clang-tidy-modern.sh:2]: Working my way up PCL's [dependency graph](https://user-images.githubusercontent.com/1709142/87282880-fd5cc080-c4f4-11ea-878e-9c7de9c71eb8.png), this PR hasn't gotten to the `kiss_fft` level.

File: .clang-tidy (hunk 3 above)
Comment (gnawme) [reply .clang-tidy:2]: Not as far as I know

File: .github/workflows/tidy-modern.yml (hunk 4 above)
Comment (gnawme) [reply .github/workflows/tidy-modern.yml:16]: I seem to remember that something in the Github Workflow had a dependency on git

File: .github/workflows/clang-tidy-modern.sh (hunk 2 above)
Comment (kunaltyagi) [reply .github/workflows/clang-tidy-modern.sh:2]: In that case, I'd recommend adding a comment like:
```bash
# ignoring kiss_fft, leaving it for future PR. Tracked in Issue #<issue here>
```

File: .github/workflows/tidy-modern.yml (hunk 5 above)
Comment (gnawme) [reply .github/workflows/tidy-modern.yml:29]: Is env:20.04 the new baseline for CI?

Comment (gnawme) [reply .github/workflows/tidy-modern.yml:29]: Apparently, 20.04 is not rolled out yet:

> Ubuntu-latest workflows will use Ubuntu-20.04 soon. For more details, see https://github.com/actions/virtual-environments/issues/1816

**Comment by SunBlack** [comment]
Just note: [Here](https://github.com/PointCloudLibrary/pcl/pull/2728#issuecomment-448958612) was the decision not to add `override` to destructors.

**Comment by kunaltyagi** [comment]
That's true. But the CppCoreGuidelines reaffirmed their position https://github.com/isocpp/CppCoreGuidelines/pull/1448 (in 2019, that discussion is from 2018, and @taketwo says the Core Guidelines position is prefered).

What are your thoughts on this @SunBlack ?

**Comment by SunBlack** [comment]
We are using `override` for dtor in our project sine years ;-). I prefer it as on this way you can be sure, that the `virtual` is not missing in the base class. Just wanted to mention it that there was the decision as I didn't knew CppCoreGuidelines changed their mind.

I just would prefer to have these modernizations in separate PRs like in #2731, as we can apply them to the whole project once. And this PR should be for enabling the checks on the CI. Imho we should enable this not just for some modules first. Instead for the whole project and than add fixed categories one by one (we just have to exclude 3rd party code somehow - maybe move the 3rd party code to a separate dir in the root project?)

**Comment by kunaltyagi** [comment]
> And this PR should be for enabling the checks on the CI

To confirm, you are advising splitting this PR into 2 parts:
* switch on the github workflow for clang-tidy checking (and the relevant per-module changes to cmake)
* apply mitigations (one or several)

> we just have to exclude 3rd party code somehow - maybe move the 3rd party code to a separate dir in the root project

If code reorganization is needed, it'd get quite messy. It might be better to have multiple CI workflows, each with their "exception"-list and config till we manage to bring order. If we delay clang-tidy due to code re-organization, I fear we might not end up with either of them. Moreover, the piece-meal approach will give us insight into any commonality between the workflows and help in re-organization

**Comment by SunBlack** [comment]
> To confirm, you are advising splitting this PR into 2 parts:
> 
> * switch on the github workflow for clang-tidy checking (and the relevant per-module changes to cmake)
> * apply mitigations (one or several)

Yep as otherwise we end like with clang-format: Some parts of the project are modernized, other not. As already most important changes are done via #2731 this should be mostly only smaller PRs now in preparation for including it in the CI.

> > we just have to exclude 3rd party code somehow - maybe move the 3rd party code to a separate dir in the root project
> 
> If code reorganization is needed, it'd get quite messy. It might be better to have multiple CI workflows, each with their "exception"-list and config till we manage to bring order. If we delay clang-tidy due to code re-organization, I fear we might not end up with either of them. Moreover, the piece-meal approach will give us insight into any commonality between the workflows and help in re-organization

Indeed. Maybe it works to add `set(CMAKE_EXPORT_COMPILE_COMMANDS ON)` in the root and `et(CMAKE_EXPORT_COMPILE_COMMANDS OFF)` just in the 3rd party projects? Or isn't it possible to turn off it just for some projects? I don't know how the integration on GitHub really works: In our project we run clang-tidy on all code and filter it later (yeah from the performance to exclude it before would be better, but it is easier as the Jenkins-Plugin handles it ;-) )

**Comment by kunaltyagi** [comment]
Let's get the low hanging fruits first :)

Could you take the lead on the first 2 parts of "get clang-tidy warnings, deduplicate them and than format them into a format Azure understand" given your experience (or if the project is open source, link the jenkins file for reference)?

@gnawme Sorry for all the chatter, I'll summarize the actionable items for you :). Based on @SunBlack's recommendation, please split the PR in two parts:
* Add empty cmake changes to all modules + clang-tidy with no checks
* Add a config + changes

**Comment by SunBlack** [comment]
> Could you take the lead on the first 2 parts of "get clang-tidy warnings, deduplicate them and than format them into a format Azure understand" given your experience (or if the project is open source, link the jenkins file for reference)?

The project we are using is [Jenkins Warnings Next Generation Plugin](https://github.com/jenkinsci/warnings-ng-plugin) and the file it can parses from Clang-Tidy are like [here](https://github.com/jenkinsci/analysis-model/blob/master/src/test/resources/edu/hm/hafner/analysis/parser/ClangTidy.txt). Not sure: You can tell `run-clang-tidy` to print files as `yml` - maybe they are better for converting into a format Azure can. Not sure which [formats](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/test/publish-test-results?view=azure-devops&tabs=trx%2Cyaml) fits our need at best:  "CTest, JUnit (including PHPUnit), NUnit 2, NUnit 3, Visual Studio Test (TRX), and xUnit 2"

**Comment by gnawme** [comment]
> Let's get the low hanging fruits first :)
> 
> Could you take the lead on the first 2 parts of "get clang-tidy warnings, deduplicate them and than format them into a format Azure understand" given your experience (or if the project is open source, link the jenkins file for reference)?
> 
> @gnawme Sorry for all the chatter, I'll summarize the actionable items for you :). Based on @SunBlack's recommendation, please split the PR in two parts:
> 
> * Add empty cmake changes to all modules + clang-tidy with no checks
> * Add a config + changes

@kunaltyagi @SunBlack 

- #4249 already applied `clang-tidy` to the entire project, but nobody wanted to review the PR; so this PR is already part of a divide-and-conquer strategy, which takes the approach of walking up the dependency graph
- To answer your question in an earlier discussion -- yes, you can enable/disable `clang-tidy` at the module level by adding `set(CMAKE_EXPORT_COMPILE_COMMANDS ON)` in the `CMakeLists.txt` for that module (which is the approach that this PR is taking)

**Comment by kunaltyagi** [comment]
> #4249 already applied clang-tidy to the entire project

That's not what I meant. Essentially, simply split this PR in 2 parts:
* changes purely for enabling clang-tidy
* config file for clang-tidy and changes due to addition of that file

**Comment by gnawme** [comment]
> > #4249 already applied clang-tidy to the entire project
> 
> That's not what I meant. Essentially, simply split this PR in 2 parts:
> 
> * changes purely for enabling clang-tidy
> * config file for clang-tidy and changes due to addition of that file

@kunaltyagi OK, I removed the `.github/workflows` directory so CI won't be triggered, and left the `clang-tidy` changes in place. Hope that's what you meant.

Github workflows to be restored in the future (or perhaps replaced by what @SunBlack has in mind).

**Comment by kunaltyagi** [comment]
Kinda. But in reverse.

Could you make a PR which has a github workflow for clang-tidy with no checks?
As for this PR, expand the checks to all sub-modules (if you haven't). (And no, please keep the checks fixed, don't add more checks)

File: .clang-tidy (hunk 3 above)
Comment (kunaltyagi) [inline .clang-tidy:2]: Should we add the following:
```
WarningsAsErrors: '#same content as in checks, but for errors'
FormatStyle:     file
```

**Comment by gnawme** [comment]
> Kinda. But in reverse.
> 
> Could you make a PR which has a github workflow for clang-tidy with no checks?
> As for this PR, expand the checks to all sub-modules (if you haven't). (And no, please keep the checks fixed, don't add more checks)

Sure; what are you trying to accomplish, exactly? #4249 has checks for the entire project; if I remove the GitHub workflows from the PR, will that accomplish your latter aim?

BTW, who approved `using namespace std` going back into the code?

**Comment by SunBlack** [comment]
> Sure; what are you trying to accomplish, exactly? #4249 has checks for the entire project; if I remove the GitHub workflows from the PR, will that accomplish your latter aim?

You could run clang-tidy with all checks and add all checks which doesn't show an issue already to the workflow.

**Comment by mvieth** [comment]
@gnawme I believe this pull request can be closed, since pull request https://github.com/PointCloudLibrary/pcl/pull/4629 is now merged?

**Comment by gnawme** [comment]
> @gnawme I believe this pull request can be closed, since pull request #4629 is now merged?

@mvieth Possibly; this PR went without attention for so long that I've forgotten.

**Pull request closed without merging**